src/dependency-manager/src/utils/interpolation/parameter-interpolator.ts

```diff
@@ -0,0 +1,119 @@
+      if (node instanceof ServiceNode) {
+        friendly_name_map[node.ref] = {};
+        for (const friendly_name of Object.keys(node.node_config.getDependencies())) {
+          const top_level_node = graph.nodes
+            .filter(n => n instanceof ServiceNode)
+            .map(n => n as ServiceNode)
+            .find(n => n.node_config.getName() === friendly_name);
+
+          if (top_level_node) {
+            friendly_name_map[node.ref][friendly_name] = top_level_node.namespace_ref;
+          }
+        }
+      }
+    }
+    return friendly_name_map;
+  }
+
+  public static mapGraphToInterpolationContext(graph: DependencyGraph, interface_context: EnvironmentInterfaceContext): EnvironmentInterpolationContext {
+    const environment_context: EnvironmentInterpolationContext = {};
+
+    for (const node of graph.nodes.filter(n => n instanceof ServiceNode).map(n => n as ServiceNode)) {
+      const service_context = ParameterInterpolator.mapNodeToInterpolationContext(node, interface_context[node.ref]);
+      environment_context[node.namespace_ref] = service_context;
+    }
+
+    return environment_context;
+  }
+
+  public static mapNodeToInterpolationContext(node: ServiceNode, interface_context: ServiceInterfaceContext): InterpolationContext {
+    return {
+      parameters: Object.entries(node.node_config.getParameters())
+        .reduce((result: { [key: string]: any }, [k, v]) => {
+          result[k] = v.default;
+          return result;
+        }, {}),
+      interfaces: Object.keys(node.interfaces)
+        .map(interface_name => {
+          return { key: interface_name, value: interface_context[interface_name] }
+        })
+        .reduce((result: { [key: string]: any }, { key, value }) => {
+          result[key] = value;
+          return result;
+        }, {}),
+    };
+  }
+
+  public static isNullParamValue(param_value: ParameterValueV2) {
+    return param_value === null || param_value === undefined;
+  }
+
+  public static namespaceExpressions(node_ref: string, expression_string: string, friendly_name_map: { [key: string]: string }) {
+    if (typeof expression_string !== 'string') {
+      return expression_string;
+    }
+    if (!expression_string.includes('${')) {
+      return expression_string;
+    }
+
+    let namespaced_value = expression_string;
+
+    const interfaces_search_string = /\$\{\s*interfaces/g;
+    namespaced_value = namespaced_value.replace(interfaces_search_string, `$\{ ${node_ref}.interfaces`);
+
+    const parameters_search_string = /\$\{\s*parameters/g;
+    namespaced_value = namespaced_value.replace(parameters_search_string, `$\{ ${node_ref}.parameters`);
+
+    const bracket_notation_matcher = /\$\{\s*dependencies\[(.*?)\]\./g;
+    const dot_notation_matcher = /\$\{\s*dependencies\.(.*?)\./g;
+
+    namespaced_value = namespaced_value.replace(bracket_notation_matcher, (m) => {
+      const dep = ParameterInterpolator.extract_friendly_name_from_brackets(m);
```

Comment:
Not loving this nested regex. Can we use a match group?

Reply:
I'll try to generalize this along these lines and your point about adding new supported blocks. That's a good idea.